### ISSUE TITLE
Maintaining backwards compatibility in V2

### ISSUE THREAD

**Issue by jerriep** [auth0/auth0.net#45]
I had a brief look into maintaining backwards compatibility with V1 of the .NET Client when doing V2, but in the 30 minutes I spent I found a number of issues. Here are some of them:

* Right now the Client class is used for the API client, and there seems to be no classes to represent an actual Client as described in the API (https://auth0.com/docs/api/v2#!/Clients/get_clients). So I would like to change the Client class to ApiClient or Auth0Client or something like that, so we can use Client to represent an actual client.

* Send verification email is not available in V2, so needs to be removed.

* The updating of user metadata, password and email has been consolidated into the PATCH request on the users resource, so it can really be removed.

* When creating a connection, it seems the concept of a provisioning ticket has been removed.

* Getting users by connection is not available anymore.

Spending more time and going into more detail, I can probably find some more issues, so my question is:

How important is it to maintain backwards compatibility? 

We can attempt to do it as far as possible, but in a number of cases it is going to be basically impossible. Would it therefore not be better to make a clean break while we have the chance?

The big concern would be for people using the Nuget package and upgrading. Their code is going to break horribly. We can perhaps create new, separate Nuget package for V2 to make it more clear that it is a completely new version?

Also, the coverage of the existing API is not too great. It seems many endpoints such as `/clients`, `/device-credentials`, `/rules`, `/blacklists` etc is not covered. When we start adding these, the number of methods of the base `Client` class (or ApiClient in new version) is going to be long.

So I want to propose another change as far as this is concerned. I want to propose a style similar to what the GitHub people did for the Octokit.net library (https://github.com/octokit/octokit.net)

They are basically grouping the related methods together under sub-clients which make it much easier to find methods you are looking for. Here is the basic definition of their `IGitHubClient` interface:

```
public interface IGitHubClient
    {
        IConnection Connection { get; }

        IAuthorizationsClient Authorization { get; }

        IActivitiesClient Activity { get; }

        IIssuesClient Issue { get; }

        IMiscellaneousClient Miscellaneous { get; }

        IOauthClient Oauth { get; }

        IOrganizationsClient Organization { get; }

        IPullRequestsClient PullRequest { get; }

        IRepositoriesClient Repository { get; }

        IGistsClient Gist { get; }

        IReleasesClient Release { get; }

        ISshKeysClient SshKey { get; }

        IUsersClient User { get; }

        INotificationsClient Notification { get; }

        IGitDatabaseClient GitDatabase { get; }

        ISearchClient Search { get; }
    }
```

So if we can do something similar it would group all the related methods together nicely and line up exactly with your API documentation. So if a user want to get all clients, they would type something like:

```
var apiClient = new ApiClient(....);
var clients = apiClient.Client.GetAll();
```

or if they want to create a user, they can write something similar to this:

```
var client = new ApiClient(...);
var user = new NewUser
{
   Connection = "....",
   Email = "....",
   // etc....
}
apiClient.Users.Create(user);
```

Obviously this new layout is also going to break backward compatibility horribly, but I think the final result will be better.

Would love to hear your feedback on this.

**Comment by sandrinodimattia** [comment]
So, first I don't think we should maintain backwards compatibility at all. APIv2 is our opportunity of doing something better with our client for .NET. I would not create a new NuGet package for this, simply apply semver and change the version from 1.4.0 to 2.0.0

The GitHub client for .NET is a great design for an API client, so nothing agains using that same design for our own API client. Now I think we will still need 2 "root" clients in our library:

 - ApiClient, which talks to API v2 - this uses JWT tokens for authentication
 - AuthentciationApiClient or AuthenticationClient, which talks to our Authentication Endpoints (https://auth0.com/docs/auth-api) - this uses Client ID/Client Secret for the token exchange endpoint

Now in addition to that, I think we should take the following into account when designing the next API client:

 [ ] Support for .NET 3.5/4/5 + CoreCLR + Universal Apps + Xamarin (like AutoMapper for example: https://github.com/AutoMapper/AutoMapper/tree/develop/src)
 [ ] Expose both synchronous and asynchronous methods (Tasks + async/await)
 [ ] Retry policies
 [ ] Better logging support (pluggable, ETW, ...)
 [ ] Expose a request context with additional information (like how long the request took)
 [ ] Better error handling
 [ ] Test Coverage
 [ ] Integration tests (should test auth0.com and private instances, because in API v2 there are a few differences when running on auth0.com or in a private instance)
 [ ] Automated build, test and push to NuGet (Travis/AppVeyor/MyGet)
 [ ] SemVer
 [ ] Automatic generation of the client where possible (Swagger)

Robert already did some work on a new client for .NET, so before starting anything let's review his implementation and reuse where possible.

**Comment by jerriep** [comment]
Great, I will wait for Robert's feedback. Can we tag him in this?

As far as having both async and non-async methods; is there a specific reason you want non-async methods? 

HttpClient, which I believe is the correct library to use if we want full PCL compatibility, does not support non-async methods.

Furthermore I was speaking with @ntotten to use Refit (https://github.com/paulcbetts/refit) for the wrapper around the raw API calls, and that also does not support non-async since it sits on top of HttpClient.

**Comment by ntotten** [comment]
@advancedrei Any feedback on this?

**Comment by advancedrei** [comment]
Thanks for including me in this process. I'm genuinely honored that you guys value my opinion, and I'm happy to be a part of this.

So there's a lot that was brought up in this thread, but I'll try to tackle as much as I can. I have done quite a bit of work prototyping a new version, and I'll cover that towards the end.

**Backwards compatibility:** That's not going to be possible IMHO. The design of your APIs are totally different. If we SemVer to 2.0, we can break stuff. This is an opportunity to start fresh, so we should take it.

**Different NuGet Package:** I agree that we need new packages, mostly because the old ones have no coordinated structure or naming conventions, they are just a hodgepodge of code. The old NuGet packages need to be updated with descriptions that point to the existence of the new ones.

**Sync + Async:** @jerriep is correct, HttpClient is async-only, so we should put the focus on having rock-solid Async support.

**OctoKit Style:** I agree that functional grouping around the client is very helpful/useful to the developer, and that kind of style should definitely be considered (but probably only actually executed after the functional groupings make themselves apparent).

**Structured for Platforms:** ABSOLUTELY. The prototype I put together is specifically structured for this.

**Post from @sandrinodimattia:** I agree with most of the design goals, although automated builds, logging, diagnostic headers, etc. should all come after the core architecture has been completed. I disagree, however, with the structure of the "roots" of the system, which I will touch on in the next bullet point..

**My Personal Goals:** I've spent countless hours working with, and making improvements on, Auth0's existing NuGet packages. There is some great stuff in there, but it's really a hassle to get your head around what you need, different packages do things like processing Claims differently, and it can be frustrating. I've made improvements to the Auth0 implementations without being able to push them back into what you guys ship, so for me, this is my opportunity to have everyone benefit.

**My goal is to put the "Kit" back in SDK.** This project is should not be about building an API client. It should be about making Auth0 a joy to use for the .NET developer, by having a complete, integrated SYSTEM for implementing auth in any app that is both intuitive and well-documented. IMHO, everything comes second to that.

https://github.com/advancedrei/Auth0.Portable is my attempt at a prototype for that goal. It combines everything for .NET into a single solution and repo, because it's all part of one system. It understands that you shouldn't have to import HttpClient into your app if you just want to store an Auth0User in your database. it understands that you should get the exact same *standardized* Claims back if you call the handler from an OWIN app, a WebAPI app, or a Windows 10 app.

The Auth0.Portable project is built using PortableRest, which is a replacement for RestSharp, built on HttpClient, and is similar to Refit. Refit is nice when there is a solid 1-1 mapping between the API and it's implementing client, but Refit assumes a lot about the developer's skill level. For example, search for "LogInAndCallApi" on the Refit Readme page. That kind of thinking is just too much to ask a customer to do, IMO. LendingTree.com uses it to power their "three quotes" process, and has contributed tons of improvements, and has served us well. Whether it's right for this project remains to be seen.

At any rate, AdvancedREI is already using this code in production for getting DelegatedTokens on our internal admin site, and that has helped me find weaknesses in the design and help make it better. I'm hoping to get to the ASPNET stuff soon so we can swap the implementation out and see what happens.

I think it might be a good idea at some point to have a Skype/Skype For Business/Hangouts/whatever session, so I can walk you guys through the prototype, we can talk design, and figure out a way forward.

Thoughts?

**Comment by ntotten** [comment]
This is great. Thanks @advancedrei. 

I completely agree about the structure of the .NET nugets and project. I think we should step back and look at all of our .NET libraries. I am completely in favor of getting rid of all the specific NuGet packages like `Auth0.Windows.Cs`. I think we can use the platform features of NuGet to move most of these into a single `Auth0` package.

Here are the repos we currently have:
https://github.com/auth0/auth0.net
https://github.com/auth0/Auth0.WindowsPhone
https://github.com/auth0/Auth0.Windows8.Cs
https://github.com/auth0/Auth0.Windows8.Js
https://github.com/auth0/auth0-aspnet-owin
https://github.com/auth0/auth0-aspnet5
https://github.com/auth0/auth0-aspnet-mvc4-net45
https://github.com/auth0/auth0-aspnet-apikeys-sample
https://github.com/auth0/auth0-aspnet
https://github.com/auth0/auth0-aspnet-mvc3-sample
https://github.com/auth0/DotNetOpenAuth-Auth0

I like the approach that @advancedrei took in his repo where he has a single repo with the core platform stuff. We may want to do something like having the single core repo, then have 1 for asp.net, and another for windows app. We could also move all the samples into the appropriate repo's /examples folder - which is what we are trying to do across our projects.

**Comment by ntotten** [comment]
I think that we can basically get our nuget packages down to:

`Auth0`
`Auth0.Web`
`Auth0.Web.Owin`
`Auth0.Windows`

**Comment by advancedrei** [comment]
Maybe. I don't know if it will be able to be that simple. It all depends on what makes sense from a dependency standpoint.

**Comment by ntotten** [comment]
:+1:

**Comment by ntotten** [comment]
@jerriep Is in Thailand so I'll let him suggest a time first.

**Comment by jerriep** [comment]
I am on the time zone UTC+7, and available anytime from let's say 6:30 AM until around midnight. I do not have any other meeting set up for the next couple of days, so anytime you suggest in that time range should be fine.

**Comment by jerriep** [comment]
> Refit. Refit is nice when there is a solid 1-1 mapping between the API and it's implementing client, but Refit assumes a lot about the developer's skill level. For example, search for "LogInAndCallApi" on the Refit Readme page.

Yeah this is a valid point, but remember that the use of Refit will be hidden from the end-user of our library. It is something I will need to concern myself with (as the author of the library), but they would not even know about it.

I have not used PortableRest before but I am completely open to using that as well, instead of Refit. All I am looking for is something which makes it easier to wrap the actual REST calls. I have liked Refit on my own projects before, because of its interface based approach which makes writing mocks for unit testing easier.

**Comment by advancedrei** [comment]
How about 9am PST Monday morning? That would be noon EST and 11pm ICT?

**Comment by jerriep** [comment]
Yes, this will be fine for me

**Comment by sandrinodimattia** [comment]
Hey guys, please also take time to think about SharePoint :)

Our Claims Provider is now using `auth0.net` to call API v1. Maybe it's enough to have it use the new libraries, but we still have customers on SP2010 (.NET 3.5), so we need to find a good solution for them too.

**Comment by advancedrei** [comment]
I don't know if that is going to be possible with a new HttpClient-based codebase. There may be a possibility to create a separate, self-contained version just for that, but there may need to be some investigation around what is holding them back from upgrading to newer versions of SharePoint, what features they want to use from the new APIs, etc. If only 5 clients are on SP2010, then I don't think it makes sense to target them from a business perspective.

Having said that, the structure I proposed in the prototype makes it a possibility to re-use the same core code and just implement a legacy Rest Framework. So it's not entirely out of the question.

**Comment by sandrinodimattia** [comment]
> Having said that, the structure I proposed in the prototype makes it a possibility to re-use the same core code and just implement a legacy Rest Framework. So it's not entirely out of the question.

I think that would be the best approach. It's not only a SharePoint problem, we might also have customers that are just on .NET 3.5 with ASP.NET Web Forms. So if we could give them a way to use something like RestSharp instead of the new HttpClient base that would be great.

And about the async part, I think we should still support thin sync wrappers around the async methods for our customers who cannot use async/Task based methods (like here: https://github.com/Azure/azure-storage-net/blob/40c04ce1dbbc0c569316da0e3fe47b9bff5d3042/Lib/ClassLibraryCommon/File/CloudFile.cs#L976-L984)

**Comment by advancedrei** [comment]
RE: Sharepoint, etc - I'm not usually a huge fan of Interfaces outside of DI scenarios, mostly because I think people spend too much time testing the wrong things, and not enough time testing others, or too much time making unnecessary things pluggable. But in this case, an IAuth0RestClient interface would let us make sure people have a seamless upgrade path between Framework versions regardless of the underlying transport implementation. We focus on building an outstanding modern version, and then you guys could backport it to RestSharp on 3.5 afterward.

RE: Async - So I'm curious to see how something like that would potentially play out. I'm not an Async expert by any means, and I have a particular section of our codebase where we cannot await methods because of the Reactive Extensions' IObserver pattern. So we just call SomeAsyncMethod().Result or SomeAsyncMethod().Wait(). It feels kind of dirty when I have to do that, though, so I'm curious if there is an alternative.

**Comment by jerriep** [comment]
Great, we can talk this through in more detail on Monday. How do you want to do the call on Monday @advancedrei ?

**Comment by sandrinodimattia** [comment]
Can someone post a short summary of the call here?

BTW, something interesting to go through in regard to PCL/UAP: https://github.com/davidfowl/aspnetvnextwebapiapp/blob/master/Generations.md

**Comment by jerriep** [comment]
Here you go:

* As far as the Sync / Async issue is concerned, we decided to keep the library forward looking and stick with async. Using the Microsoft.Bcl.Async package, this means we can support framework versions as far back as 4.0.
* Where people want sync calls (in the more modern .NET versions), we can investigate at a later stage to make sync extension methods available. This would be opt-in through an optional Nuget package.
* For situations like SharePoint 2010 which requires .NET 3.5, we will write a purpose fit solution. This was also decided because the current SP 2010 solution requires a very limit set of the Auth0 API methods.   
* Based on Robert's feedback, we decided to split the "main" library into 3 parts. A core containing basically only the models. A library for the Management APIs, and a library for Authentication APIs. The last 2 will use the core library containing the models. All the various platform specific libraries for OWIN, ASP.NET 5, UWP etc will then make use of these 3 libraries where appropriate.
* The existing code in the Auth0.Net repo will be moved to a "legacy" branch. 
* I will develop on my fork of the repo in a new V2 branch, and then when ready submit a PR to the Auth0.Net repo
* I will also use Robert's existing work (https://github.com/advancedrei/Auth0.Portable) as a starting point for the new V2
* We confirmed that the library will need to be fully documented with XML comments. We will still investigate a final solution to generate developer documentation from that.
* Robert requested that the developer documentation for the library also contains proper sample code 

@advancedrei and @ntotten please comment if I missed anything

**Comment by advancedrei** [comment]
Yep, that covered it.

As an aside, I'm leaving this here for reference later: with ASP.NET 5, we may not need to dig that deep into an Auth0 integration: https://github.com/aspnet/Security/issues/537

**Comment by jerriep** [comment]
An early version of the new Management API is available at https://github.com/jerriep/auth0.net/tree/V2

I have only implemented the API methods to manage clients, but it should give you an idea of the general direction I am moving in. Still a lot of work to be done, but feedback is welcome if you guys are worried about the current approach.

To get an idea of how the API will be used, you can look at the sample program at https://github.com/jerriep/auth0.net/blob/V2/tests/ConsoleTestWorkbench/Program.cs

**Comment by ntotten** [comment]
@jerriep this is looking very good. I like the approach. Very clean.

**Comment by advancedrei** [comment]
So, I wasn't sure how the apiClient.Group.XXXAsync was going to work out, but I like it. Great work so far.

The only problem that I see so far is that people are going to name their `RestClient` variable `"auth0Client"` or `"client"` or something like that, and `client.Clients` is misleading/not intuitive as to what it could represent. I understand that the actual API stack is not likely to change, but I don't think that the API endpoint names for v2 are as intuitive as they could be.

I'll have some time over the weekend to do a more thorough review and see if I can poke any holes in it, but so far so good! :)

**Comment by jerriep** [comment]
I have Resharper, so renaming things are easy for me ;)

I am very open to a naming convention which makes better sense.

**Comment by advancedrei** [comment]
The `___.Clients.___` naming should match the REST endpoints. I just don't know if the endpoints have the right names. But this IS the management API, so fewer people will use it.

**Comment by jerriep** [comment]
I match it up as it is in the documentation (and API Explorer):
![image](https://cloud.githubusercontent.com/assets/1006420/11138796/881fa692-89f9-11e5-9052-05ddf0b3830b.png)

This (for the most part) matches up with the endpoints.

Was yours just a general comment of principal, or are you referring to specific instances where it does not?

**Comment by ntotten** [comment]
We should match the REST API names. We won't be changing them any time
soon. :)

**Comment by jerriep** [comment]
@advancedrei BTW, is it possible to do multipart/form-data with PortableRest? 

For example for import users job (https://auth0.com/docs/api/v2#!/Jobs/post_users_imports) I need to do that...

**Comment by jerriep** [comment]
@ntotten Sure. Like I said they match up for the most part, with the exception of the Emails one (AFAIK), where the endpoint is /emails/provider, so I'll change it to `EmailsProviderClient`

**Comment by advancedrei** [comment]
@jerriep I don't believe it does it right this second, but I'm fairly certain there is a PR for is that I can get integrated sometime this weekend.

**Comment by jerriep** [comment]
Yes indeed :)
https://github.com/advancedrei/PortableRest/pull/53

Whenever you get a chance over the next week or so would be good, thanks

**Comment by advancedrei** [comment]
Update your PortableRest to 3.1.0-Beta2 and let me know if the new functionality works for you.

Thanks!

**Comment by jerriep** [comment]
Thank you @advancedrei ! Will get back around to that particular code later today (or tomorrow). Will let you know if I have any issues

**Comment by jerriep** [comment]
@advancedrei I'm running into some issues. Will open issue on your PortableRest repo

**Comment by jerriep** [comment]
@ntotten I have reached the point in V2 where I have full coverage of the Management API. There is still a number of outstanding issues, but I want to find out whether I can do a PR and get the existing work into this repo (even if under a V2 branch), so I can start getting input from everyone?

**Comment by jerriep** [comment]
@advancedrei BTW, I have a nice new problem for you:
https://github.com/advancedrei/PortableRest/issues/58

**Comment by advancedrei** [comment]
On it.

So I'm wanting to contribute some stuff to the Core, so I can take a dependency on it and not the custom version of V1 that I currently have to maintain. I wanted to find out what the rationale is for having some objects in the root folder, and some objects under "Models".

I think there is some merit to separating objects that are core to Auth0, and objects that are specific to constructing certain API requests, like the XXXCreateRequest objects. Was that the intention?

**Comment by jerriep** [comment]
The stuff in the root are still left overs which I brought in from your initial work. The are related to the Auth API, so I left them in until I get around to actually doing the Auth API. Whatever is left after that will be cleaned out. 

The xxxCreateRequest are also models though. The thing is that the model passed when creating or updating many time contains less (or more) fields that the actual model returned when retrieving an object, so this is how I differentiate between them.

They are not used to construct API requests. They are still models.

Can you also please specify what stuff you want to contribute to the core. As I said I am still busy as I am doing the Auth API next, and I do not want us to trip over each other. If it makes sense that I add it, you can pass it along to me.

**Comment by advancedrei** [comment]
I think you are being too literal with the word "model". Any object that doesn't have methods could be considered a model, that doesn't mean we need to be explicit about burying them in a "Models" namespace. 

The way you are delineating base objects vs request-specific objects makes sense. But from an organizational standpoint, it may make sense to move anything that has the "Request" suffix, and will not be used by the AuthAPI, to the ManagementAPI library, because they are specific to that library. For example, "User" and "UserBase" are core objects that will be used in multiple situations. "UserCreateRequest" is a model whose sole purpose is to pass data to an API request, and will not be used for any other purpose.

Regarding contributions, I was looking at support for the objects that can be POSTed from the a currently-executing Rule, and where to put those objects in the Core. Since you had said you hadn't started on it yet, I had also created the AuthAPI project and was looking at how to share some of the core pipeline code you wrote between both libraries, and what technique would make the most sense. I don't want to step on your toes, that's why I was going to submit a PR for your review.

**Comment by jerriep** [comment]
OK, I got you on the Request stuff. I can certainly move those to the Management API project.

Regarding the contributions, I am also creating an Auth API project right now and starting to work on that, so if you also submit a Auth API project via another PR project we are going to trip over each other, so please just hang on a week or two until I am done with it.

Contributions to the Core project should be fine, but just know that I am working on Core models to the Auth stuff as well, so even there we can trip over each other...

**Comment by jerriep** [comment]
Is the stuff you are talking about in your https://github.com/advancedrei/Auth0.Portable project?

If so, then I can just grab it from there and integrate into the work I am doing. If you can just point out the specific items.

**Comment by advancedrei** [comment]
The stuff from there you already pulled into the main project. What I was adding was stuff from a PR that was not accepted a long time ago.

I don't want to trip you up, but there has to be a way to be able to make it work. If you could maybe think about working in smaller increments, I'd like to be able to contribute too. I have additional properties and documentation that I would like to get worked in there also.

Thanks! :)

**Comment by jerriep** [comment]
OK, let's do it this way:

* I will create a PR just now with the Auth project added, so it is available for both of us to work on.
* You can then let me know which API endpoints it is that you are working on, and I will stay clear of those. 

Can this work for you? :)

**Comment by advancedrei** [comment]
That works, thanks man. I really appreciate it. 

I wasn't really thinking about specific endpoints, just so you know. You're doing great work there, and I don't want to mess with it. I was more about how the plumbing you already wrote in the Management library could be reused to minimize duplicate code. Was also trying to work out the Windows10/.NET 4.6-related issues with PortableRest, and how the packages should be structured to avoid at least one of the issues you ran into. Microsoft really screwed the pooch with all this PCL / UWP / NetCore nonsense.

**Comment by jerriep** [comment]
Cool, thanks Robert :) It is not critical right now (I assume you have Thanksgiving responsibilities), but we will need to get it sorted in the next few weeks if we want to generate working Nuget packages.

**Comment by advancedrei** [comment]
@jerriep How did you have two active PRs at once? The last time I tried to do that, it added my later check-ins into the first PR. Thanks!

**Comment by jerriep** [comment]
You should have 2 different branches to do that. 

It the work of the second branch is however based on the work from the first branch, you may as well just keep it in one branch, because then the PR for branch 2 will also contain the same changes as the PR for branch 1.

**Comment by advancedrei** [comment]
Ah. I think I'll just wait then. If you have a chance to give your OK on the PR I sent over, I'll ping Nathan to check it out, then queue up the next one. :)

**Comment by jerriep** [comment]
I want to bring up the issue regarding documentation again. I am busy going through the project and ensuring that XML documents are everywhere.

Regarding producing a physical help file - what is the use case for this? I think very few developers will ever look at that, or even know that it is there. Or is the idea to produce documentation and then host it on a website somewhere?

if so, do we not run the danger of that documentation running out of sync with the official API documentation?

Would love to hear everyone's thoughts on this...

**Comment by jamestoyer** [comment]
Would it be possible to host the documentation using Github pages? Then on every release automatically updating that. Surely that would keep it as up to date as possible

**Comment by jamestoyer** [comment]
Also, and I guess this is a side point, is it worth this being and "ASP.Net 5" project? I mean would it be worth getting that sorted now such that while it won't support CoreCLR, it would support .Net 4.5+ and cross target more easily. This does effectively means that the pre .Net 4.5 requisite is lost. 

I only say this as I'm starting to write ASP.Net 5 web apps, and would at some point like to end up with them being cross platform

**Comment by ntotten** [comment]
@jamestoyer We will be publishing the docs to auth0.com/docs autoamtically when we push a release.

**Comment by ntotten** [comment]
@jamestoyer I don't quite understand what you mean about it being an ASP.NET 5 project. We are targeting PCL because the library is used on many different platforms like windows 10, asp.net, xamarin, etc. I am not sure what the status of CoreCLR is or exactly what we would gain by targeting it, but we can probably explore it in the future. Right now we are very close to pushing this release out so I dont think we will be making a big change at the moment.

**Comment by jamestoyer** [comment]
@ntotten What I mean is the new project structure style, whereby `.csproj` files are replace with `project.json` files that allow you to target multiple frameworks, including the [CoreCLR](https://github.com/dotnet/coreclr). 

The CoreCLR target would allow people who which to host applications on Linux, or OSX, as well as Windows. It's more about the idea of being cross platform. As far as I'm aware, and I could be wrong, without converting to the new project structure you can't target the CoreCLR.

As you're close to pushing out this release, I can just open up another issue to track this if you would like.

**Comment by advancedrei** [comment]
Having everything be portable is laying the groundwork for CoreCLR support. There are so many platforms the .NET SDK needs to support, and there was a conscious decision initially to have Auth0's SDK only run on currently-supported platforms. More work will be done later to support other stuff (like integrating with ASP.NET in general) after the core stuff gets out the door. HTH!

**Comment by ntotten** [comment]
What Robert said. :)

**Comment by jamestoyer** [comment]
Cool. I guess I'll just have to wait a little longer use Auth0 and ASP.Net 5 on Linux :stuck_out_tongue_winking_eye:

**Comment by advancedrei** [comment]
@jamestoyer It won't be too much longer. Likely very near to ASP.NET 5 RTM.

**Comment by sandrinodimattia** [comment]
And just to clarify, you can already use Auth0 for authentication in ASP.NET 5 applications (both API and MVC applications): https://github.com/auth0/auth0-aspnet5/tree/master/examples
